README
/*
 *  radiusplugin -- An OpenVPN plugin for do radius authentication 
 *					and accounting.
 * 
 *  Copyright (C) 2005 EWE TEL GmbH/Ralf Luebben <ralfluebben@gmx.de>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
 

-----------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------
The plugin was developed during my diploma thesis for the EWETEL GmbH (www.ewetel.net). 
They main idea was to create a virtualized SSL-VPN with RADIUS support. 
The virtualization was done by XEN. 
At this place I would like to thank the people from EWETEL GmbH and EWE AG (www.ewe.de) 
who supported me with the required equipment, hardware 
and helped me with all my questions.

The plugin was developed and tested under Debian Sarge and OpenVPN 2.0 on x86-machines. 
-----------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------



The RADIUS-Class:
-----------------
For this plugin was developed a RADIUS-Class in C++. It was based on the C-library FF RADIUS Library v0.1 from
Fabrizio Fiorucci (http://ffradiuslib.ff.homelinux.net).
If somebody is interested in a standalone RADIUS-class please contact me. There will be some little changes.


DOCUMENTATION:
--------------
The documentation is done with doxygen. 
To generate the documentation execute "doxygen doxygen.conf" in the folder of the plugin.


INSTALL:
--------
1. Requirements: the libgcrypt library.

2. Compile: see section COMPILE

3. Set the plugin in the OpenVPN configfile: plugin /path/to/plugin/radiusplugin.so [configfile]

4. Create a configfile for the plugin, see radiusplugin.cnf. If no path is set in the configfile of OpenVPN
	the plugin will read the file /etc/openvpn/radiusplugin.cnf: .
   The configuration is case-sensitive, for an example see radiusplugin.cnf.

5. Set some configuration in the config file of the OpenVPN server:
	- status /var/log/openvpn/status.log 1  # The status, where the plugin reads the accounting information from.

6. Configure in the config file of the OpenVPN client:
	- auth-user-pass [/etc/openvpn/passwd]  # Set this for sending a username and password to the server, this values are the username and password for the radius authentication.

7. If you use auth_control_file (>= OpenVPN 2.1 rc8), the plugin directory needs write permission to the OpenVPN directory or use the "--tmp-dir" option to define a directory for the files.

8. Maybe write your own script for vendor specific attributes (see section VENDOR SPECIFIC ATTRIBUTES).


OTHER DOCUMENTATION
-------------------
An "OpenVPN - RADIUS - MySQL Howto" in german can be found at:
http://www.roessner-net.com/


FREERADIUS EXAMPLE
-------------------
user1  Cleartext-Password := "testing"
       Service-Type = Framed-User,
       Framed-IP-Netmask = 255.255.255.0,
       Framed-IP-Address = 10.8.0.33,
       Framed-Routing = Broadcast-Listen,
       Framed-Compression = Van-Jacobsen-TCP-IP,
       Framed-Route += "192.168.101.0/26 10.8.0.1/32 1",
       Framed-Route += "192.168.111.0/24 10.8.0.1/32 1",
       Framed-Route += "192.168.112.0/24 10.8.0.1/32 1",
       Acct-Interim-Interval=5,
       Ascend-Data-Rate=100,
       Ascend-Xmit-Rate=200,
       Framed-Protocol = PPP

The Framed-Route attribute must be in the format as shown above!


COMPILE 
--------
(tested with Debian on x86, for other architectures see section TROUBLESHOOTING):

>$ make


RADIUS PACKETS and OpenVPN events (see http://openvpn.net/man.html SCRIPTING AND ENVIRONMENTAL VARIABLES)
---------------------------------------------------------------------------------------------------------
OpenVPN event                   RADIUS PACKET                                    Information
-------------                   ------------                                     ------------
auth-user-pass-verify           ACCESS REQUEST                                  parse of attributes (see below)
client-connect                  ACCOUNTING REQUEST (Start)			
-                               ACCOUNTING REQUEST (Update)                     If the user had a acct interim interval attribute the update   
                                                                                is sent periodically with this interval.(Start and stop accounting
                                                                                packets are sent ever.)
client-disconnect               ACCOUNTING REQUEST (Stop)		


RADIUS ATTRIBUTES WHICH PARSED BY THE PLUGIN AND WRITE TO THE CLIENT CONFIG FILE:
---------------------------------------------
- Framed ip address
- Acct Interim Interval
- Framed Routes (only in following format, e.g.: "192.168.101.0/26 10.8.0.1/32 1")


OTHER RADIUS ATTRIBUTES
-----------------------
- Reply Message 	the text is forwarded to stderr

VENDOR SPECIFIC ATTRIBUTES
---------------------------
If you want use vendor specific attributes the plugin can call your own program or script at the OpenVPN action
CLIENT-CONNECT and CLIENT-DISCONNECT.
If you want to use the feature you have to specify the program and a named pipe for communication
in the config file.
The file vsascript.pl shows an example.


LIMITS:
-------
commonname of the certificate 256 characters 
routes entry from the radius server 50 characters


LIMITS in RadiusConfig for the configfile:
------------------------------------------
(The limit is the length of the array minus 1!)
char serviceType[2]; 			
char framedProtocol[2]; 		
char nasPortType[2]; 			
char nasIdentifier[128]; 		
char nasIpAddress[16]; 			
char p2p[16];
char subnet[16];



TROUBLESHOOTING
---------------
- BSD:
	- RADIUS attributes FramedIP and FramedRoutes mandatory
	- RADIUS attribute FramedIP is ignored
	- Use gmake -f Makefile.bsd (maybe you have to adapt library and include path, it was tested for FreeBSD)

- Setting of Framed-Routes fails under BSD:
	- At the moment the plugin can only set the routes on Linux OSs.
	- The system call for "route ... " must be changed. Please contact me for a patch.
	- Newer version will integrate OS dependent calls. 

- OpenVPN hangs when a client connects:
	- Check OpenVPN config and check the OpenVPN log (debuglevel 7): 
		- The plugin needs following varibales from OpenVPN in ENVP-array:
			- username, password, untrusted_ip, common_name, untrusted_port, ifconfig_pool_remote_ip, verb
			- In some cases ifconfig_pool_remote_ip is missing (Depends on OpenVPN config).

- The radius packets are sent over the wrong interface:
	- Sometimes on multi interface machines with IPSec the operating system doesn't find the right interface.
	- Use the plugin from Magne J. Andreassen (it is for version 1.2a)

- The plugin writes client config files without any entry:
	- The plugin always writes the file for every connecting client, please set the ccd path 
		in the config file of the plugin to a existing dummy directory.
		
- Compiling fails on Solaris 10:
	- add following options: -lsocket -lnsl -lresolv
	
- Values for BytesIn/BytesOut are 0 in the accounting log of the RADIUS server:
	- Make sure that the status file of OpenVPN exists and the plugin can read it.
	- Use version 1 for the status file.

- Error at compiling: not initialised shared memory before the call to the MD5 function:
	- Make sure that you use libgrypt 1.2.0 or higher.

- Error at compiling: DBG: md_enable: algorithm -4195948 not available:
	- Make sure that you use libgrypt 1.2.0 or higher.

- The plugin can't write the auth_control_file.
        - The plugin needs write permission in the OpenVPN directory.



FOR TESTING
-----------

Attention: Some values are hard coded in main.cpp testing, so you can get unexpected values when testing.
For example the framedip is always 10.8.0.100, because it is read from the environment variable "ifconfig_pool_remote_ip", which is passed 
to the plugin from OpenVPN!
All hard coded variabled:
env1[0]="username=user1";
env1[1]="password=testing";
env1[2]="verb=10";
env1[3]="untrusted_ip=127.0.0.1";
env1[4]="common_name=R-VPNGateway1";
env1[5]="trusted_ip=127.0.0.1";
env1[6]="ifconfig_pool_remote_ip=10.8.0.100";

compile for test with a main function:
g++ -Wall -o main AccountingProcess.cpp Exception.cpp PluginContext.cpp UserAuth.cpp AcctScheduler.cpp IpcSocket.cpp radiusplugin.cpp User.cpp AuthenticationProcess.cpp main.cpp UserAcct.cpp UserPlugin.cpp Config.cpp RadiusClass/RadiusAttribute.cpp RadiusClass/RadiusPacket.cpp RadiusClass/RadiusConfig.cpp RadiusClass/RadiusServer.cpp  RadiusClass/RadiusVendorSpecificAttribute.cpp -lgcrypt


ChangeLog
radiusplugin_v1.1a:
- Standard configfile: /etc/openvon/radiusplugin.cnf.
- #include <unistd.h> in IpcSocket.h, is needed for compiling on some systems. 
- Correct README: A configfile must set with "-c configfile".
- Correct fprintf statement in UserAuth.cpp (line 300).
- Set debug level from 7 to 5.

radiusplugin_v1.1.a (5.1.06):
- correct calculate of ipaddress for ifconfig-push in UserAuth::createCcdFile()
- FramedIpAddress-Attribute is set to the IP address OpenVPN is assigned to the client,
	the address is read from ENVP-Array with name ifconfig_pool_remote_ip and set to 
	FramedIpAddress-Attribute at client connect. (radiusplugin.cpp, openvpn_plugin_func_v1(), part: CLIENT_CONNECT)
- suppert for topology option (OpenVPN 2.0.1)
	- read from configfile parameters subnet, p2p
	- add parameters in config with getters, setters and modify constructors and destructor
	- modify UserAuth::createCcdFile to set the right topology option

radiusplugin_1.2:
- use libgcrypt instead openssl (for gpl compability)
- checking of the authenticator field from received packets
- correct error on deleting users without an acct-interim-interval in AcctScheduler.cpp::findUser()
- send message to foreground process if no user was found (else the plugin hangs)

radiusplugin_1.2a:
- correct error if a user connects again, if he is still known by the plugin, now the user can't reconnect if he is known by the plugin
- Change GPL text
- add COPYING file

radiusplugin_1.2b:
- correct error: don't disconnect at rekeying/renegotiation

radiusplugin_1.2c:
- new algorithm to generate Acct-Session-ID, so it should be unique ever, see createSessionID in radiusplugin.cpp

radiusplugin_1.2d:
- option in OpenVPN config file is now: "plugin /etc/openvpn/radiusplugin.so [configfile]
- correct bug: close of configfile was missing

radiusplugin_1.2e:
- use RadiusClass_v1.1a for more machine independance (big endian/little endian)
- use send()/recvInt() instead of sendCode()/recvCode() for internal socket communication

radiusplugin_2.0:
- new features:
	- support of vendor specific attributes with example perl script 
	- support for OpenVPN options: dupliate-cn, client-cert-not-required, username-as-commonname
    - comments allowd in config file
    - parsing of OpenVPN config file for options (see radiusplugin.cnf)
- fixed bugs:
	- crash on rekeying if no radius server respond
	- framed ip is sent in access request packets if the client already has one (at rekeying), for dynamical key assignment through the radius server, so it will not send a new one
	- passwords and sharedsecret are shown as ****** in the logs

radiusplugin_2.0a:
- fix error in IpcSocket.h, 
  bad function name: void IpcSocket::recvBuf(User *) -> void recvBuf(User *);

radiusplugin_2.0b:
- fixed datatypes from int to uint32_t/uint64_t to prevent overflows above 2^31
- added support for gigawords
- added a Makefile
- correct bug in parsing config files
- add new config variable overwriteccfiles, allows the plugin to overwrite client config files if set to true
- update some documentation
- correct parsing of ip address in RadiusAttribute.cpp in setValue for data type IPADDRESS
- update example config file
- use string for config parsing
- correct bug in Config.cpp: parameter status was detected wrong if parameter status-version parameter is used in OpenVPN config
- don't delete client config file at CLIENT-DISCONNECT it could be already a new file from a new AUTH-USER-PASS-VERIFY
- Fix segmentation fault in radiusplugin.cpp. The error occurs if an accounting request fails after a successful authentication.

radiusplugin_2.0c:
- add IPv6 support (IPv6 patch for OpenVPN openvpn-2.1-udp6.patch)
- check if string/buffer length is zero in IpcSocket
- delete NAS port if authentication fails

radiusplugin_2.0d_beta:
- add some headerfiles to avoid compiling errors on Fedora with gcc4
- close socket in radiuspacket.cpp on retries
- add parenthesize to avoid compiler warnings 

radiusplugin_2.1:
- Support of OPENVPN_PLUGIN_FUNC_DEFERRED:
*** The communication to the radius server for authentication is outsourced in a thread.
*** If an auth control file is defined and "useauthcontrolfile=true" is defined in the config file of plugin the authentication is done the background.
*** The OpenVPN process needs write permission in the OpenVPN directory.
- Timestamps are included in the debug information.
- Adapted to OPENVPN_PLUGIN_VERSION 2
- Internal key is now based on untrusted_ip and untrusted_before.
- Separated key for status file is added.
- Update password and username if the user (key) is already known. Before if the ip and port were the same and the timeout hasn't occurred the login will success. The update prevents against this behaviour. 
- Initialization of gcrypt library added.
- Rewrite ACF file when the user is authenticated.
- Include nested configfiles from main configfile.
- Add "\r" in the configfile parser, before the carriage return was ignored.
- Add support for RADIUS reply message, the output is sent to stderr 
- Only create the client-config file, if one of attributes for it is defined.
- Add NAS port number to the session id, it's definitely unique
- Acct-Session-Id added to Access-Request packet (RFC2866)

radiusplugin_2.1a:
- Implement accounting only feature (option: accountingonly, default false)
- Implement non fatal accounting (failures during accounting let the user still connect) (nonfatalaccounting)

main.cpp
/*
 *  radiusplugin -- An OpenVPN plugin for do radius authentication 
 *					and accounting.
 * 
 *  Copyright (C) 2005 EWE TEL GmbH/Ralf Luebben <ralfluebben@gmx.de>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
 
//#include "openvpn-plugin.h"
#include "RadiusClass/RadiusPacket.h"
#include "radiusplugin.h"

#include <stdlib.h>
#include <stdio.h>


/** Only for testing the plugin!*/

/* Testcases: 
1) AUTH_USER_PASS_VERIFY, CLIENT_CONNECT, CLIENT_DISCONNECT when auth_control_file is not specified
2) AUTH_USER_PASS_VERIFY, CLIENT_CONNECT, CLIENT_DISCONNECT when auth_control_file is specified
    This testcase is more difficult:
      CLIENT_CONNECT can occur before AUTH_USER_PASS_VERIFY finishes, normally OpenVPN should take of this.
      AUTH_USER_PASS_VERIFY is called again for the same user during the first authentication happens.
3) The radius server doesn't respond: 1) & 2)
4) Slow response of the radius server (add delay sudo tc qdisc add dev [$iftoradius] root netem delay 300ms)
*/
int main (void)
{
	openvpn_plugin_handle_t context;
	//success* variables save return values of the functions
	int success1,success2,success3,success4,success5,success6,success7,success8,success9,success10,success11,success12;
	unsigned int type_mask=0;


	const char *env1[10];
	const char *env2[10];
	const char *env3[10];
	const char *env4[10];

	
	const char *argv[3];
	
	
	env1[0]="username=user1";
	env1[1]="password=testing";
	env1[2]="verb=10";
	env1[3]="untrusted_ip=127.0.0.1";
	env1[4]="common_name=R-VPNGateway1";
	env1[5]="trusted_ip=127.0.0.1";
	env1[6]="ifconfig_pool_remote_ip=10.8.0.100";
	env1[7]="untrusted_port=111";
        //env1[8]=NULL;
        env1[8]="auth_control_file=./acfuser1";
	env1[9]=NULL;
	
	env2[0]="username=user2";
	env2[1]="password=testing";
	env2[2]="verb=10";
	env2[3]="untrusted_ip=127.0.0.1";
	env2[4]="common_name=R-VPNGateway2";
	env2[5]="trusted_ip=127.0.0.1";
	env2[6]="ifconfig_pool_remote_ip=10.8.0.100";
	env2[7]="untrusted_port=111";
        //env2[8]=NULL;
	env2[8]="auth_control_file=./acfuser2";
	env2[9]=NULL;
	
	env3[0]="username=user3";
	env3[1]="password=testing";
	env3[2]="verb=10";
	env3[3]="untrusted_ip=127.0.0.1";
	env3[4]="common_name=R-VPNGateway3";
	env3[5]="trusted_ip=127.0.0.1";
	env3[6]="ifconfig_pool_remote_ip=10.8.0.100";
	env3[7]="untrusted_port=111";
        //env3[8]=NULL;
	env3[8]="auth_control_file=./acfuser3";
	env3[9]=NULL;
	
	env4[0]="username=user4";
	env4[1]="password=testing";
	env4[2]="verb=10";
	env4[3]="untrusted_ip=127.0.0.1";
	env4[4]="common_name=R-VPNGateway4";
	env4[5]="trusted_ip=127.0.0.1";
	env4[6]="ifconfig_pool_remote_ip=10.8.0.100";
	env4[7]="untrusted_port=111";
        //env4[8]=NULL;
	env4[8]="auth_control_file=./acfuser4";
	env4[9]=NULL;
	
	argv[0]="radiusplugin.so";
	argv[1]="./radiusplugin.cnf";
	argv[2]=NULL;
        system("rm acfuser*"); //remove the acf files
        
	context=openvpn_plugin_open_v2(&type_mask,(const char **) argv,(const char **) env1, NULL);
	if (context==NULL)
	{
		return -1;	
	}
	//AUTH
	//env1
	success1=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY, (const char **) argv, (const char **)env1, NULL, NULL);
	//sleep(1);
	//env2
	success2=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY, (const char **) argv, (const char **)env2, NULL, NULL);
	//sleep(1);
	//env3
	success3=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY, (const char **) argv, (const char **)env3, NULL, NULL);
	//sleep(1);
	//env4
	success4=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY, (const char **) argv, (const char **)env4, NULL, NULL);
	//sleep(1);
	
	if(success1 == OPENVPN_PLUGIN_FUNC_DEFERRED || success1 == OPENVPN_PLUGIN_FUNC_DEFERRED || success1 == OPENVPN_PLUGIN_FUNC_DEFERRED || success1 == OPENVPN_PLUGIN_FUNC_DEFERRED)
        {
          char c1,c2,c3,c4;
          ifstream file1(get_env("auth_control_file", env1));
          ifstream file2(get_env("auth_control_file", env2));
          ifstream file3(get_env("auth_control_file", env3));
          ifstream file4(get_env("auth_control_file", env4));
          do
          {
            if(file1.is_open())
            {
              file1.read (&c1, 1);
            }
            else file1.open(get_env("auth_control_file", env1));
            
            if(file2.is_open())
            {
              file2.read (&c2, 1);
            }
            else file2.open(get_env("auth_control_file", env2));
            if(file3.is_open())
            {
              file3.read (&c3, 1);
            }
            else file3.open(get_env("auth_control_file", env3));
            if(file4.is_open())
            {
              file4.read(&c4, 1);
            }
            else file4.open(get_env("auth_control_file", env4));
            sleep(1);
          }
          while(!file1.is_open() || !file2.is_open() || !file3.is_open() || !file4.is_open());
          success1=c1-'0';
          success2=c1-'0';
          success3=c1-'0';
          success4=c1-'0';
          success1=success2=success3=success4= success1 & success2 & success3 & success4;
          file1.close(); 
          file2.close();
          file3.close();
          file4.close();
          system("rm acfuser*"); //remove the acf files
        }
	
	//CONNECT
	if (success1==0 && success2==0 && success3==0 && success4==0)
	{ 
		success5=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_CLIENT_CONNECT, (const char **) argv, (const char **)env1, NULL, NULL);
		//sleep(1);
		success6=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_CLIENT_CONNECT, (const char **) argv, (const char **)env2, NULL, NULL);
		//sleep(1);
		success7=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_CLIENT_CONNECT, (const char **) argv, (const char **)env3, NULL, NULL);
		//sleep(1);
		success8=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_CLIENT_CONNECT, (const char **) argv, (const char **)env4, NULL, NULL);
		//sleep(10);
	}
	
	//cerr << getTime() << "\n\nTurn off radius\n\n";
        
	//rekeying AUTH
	//env1
        success1=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY, (const char **) argv, (const char **)env1, NULL, NULL);
		//sleep(1);
		//env2
        success2=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY, (const char **) argv, (const char **)env2, NULL, NULL);
		//sleep(1);
		//env3
	success3=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY, (const char **) argv, (const char **)env3, NULL, NULL);
		//sleep(1);
		//env4
	success4=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY, (const char **) argv, (const char **)env4, NULL, NULL);
		//sleep(1);
	if(success1 == OPENVPN_PLUGIN_FUNC_DEFERRED || success1 == OPENVPN_PLUGIN_FUNC_DEFERRED || success1 == OPENVPN_PLUGIN_FUNC_DEFERRED || success1 == OPENVPN_PLUGIN_FUNC_DEFERRED)
        {
          char c1,c2,c3,c4;
          ifstream file1(get_env("auth_control_file", env1));
          ifstream file2(get_env("auth_control_file", env2));
          ifstream file3(get_env("auth_control_file", env3));
          ifstream file4(get_env("auth_control_file", env4));
          do
          {
            if(file1.is_open())
            {
              file1.read (&c1, 1);
            }
            else file1.open(get_env("auth_control_file", env1));
            
            if(file2.is_open())
            {
              file2.read (&c2, 1);
            }
            else file2.open(get_env("auth_control_file", env2));
            if(file3.is_open())
            {
              file3.read (&c3, 1);
            }
            else file3.open(get_env("auth_control_file", env3));
            if(file4.is_open())
            {
              file4.read(&c4, 1);
            }
            else file4.open(get_env("auth_control_file", env4));
            sleep(1);
          }
          while(!file1.is_open() || !file2.is_open() || !file3.is_open() || !file4.is_open());
          success1=c1-'0';
          success2=c1-'0';
          success3=c1-'0';
          success4=c1-'0';
          success1=success2=success3=success4= success1 & success2 & success3 & success4;
          file1.close(); 
          file2.close();
          file3.close();
          file4.close();
          system("rm acfuser*"); //remove the acf files
        }

	if (success5==0 && success6==0 && success7==0 && success8==0)
	{
		//DISCONNECT
		success9=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_CLIENT_DISCONNECT, (const char **) argv, (const char **)env1, NULL, NULL);
		//sleep(1);
		success10=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_CLIENT_DISCONNECT, (const char **) argv, (const char **)env2, NULL, NULL);
		//sleep(1);
		success11=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_CLIENT_DISCONNECT, (const char **) argv, (const char **)env3, NULL, NULL);
		//sleep(1);
		success12=openvpn_plugin_func_v2(context, OPENVPN_PLUGIN_CLIENT_DISCONNECT, (const char **) argv, (const char **)env4, NULL, NULL);
	}
	
	openvpn_plugin_close_v1(context);
	if (success1==0 && success2==0 && success3==0 && success4==0 && success5==0 && success6==0 && success7==0 && success8==0 && success9==0 && success10==0 && success11==0 && success12==0 )
	{
		cerr << getTime() <<"AUTH_OK\n";
	}
	else
	{
		cerr << getTime() <<"FALSE\n";
	}

	
	return 0;
}
